### PR TITLE
Add blog post on statistical correctness

### PR DESCRIPTION
**Pull request by Bisaloo**
To preview this blog post, you need to install quarto and run `quarto preview` in your terminal.

### CODE REVIEW

File: posts/statistical-correctness/index.qmd [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  
+})
+```
+:::
+
+# Compare to a theoretical upper or lower bound
+
+An alternative strategy is to compare your result to theoretical upper or lower bound. This offers a weaker guarantee that your implementation and your results are correct but it can still allow you to detect important mistakes.
+
+You can see a [real-life example of this kind of test in the `{finalsize}` R package](https://github.com/epiverse-trace/finalsize/blob/a710767b38a9242f15ab4dcf18b02fb5b0bcf24f/tests/testthat/test-newton_solver_vary_r0.R#L1-L13). `{finalsize}` computes the final proportion of infected in a heterogeneous population according to an SIR model. Theory predicts that the number of infections is maximal in a well-mixed population:
```
Comment (pratikunterwegs) [inline posts/statistical-correctness/index.qmd:88]: Suggest moving this Epiverse example after the centroid example

File: posts/statistical-correctness/index.qmd [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  
+  y <- x
+  y[[1]] <- y[[1]] + 1 
+
+  expect_true(all(centroid(x) < centroid(y)))
+  
+})
+```
+:::
+
```
Comment (pratikunterwegs) [inline posts/statistical-correctness/index.qmd:141]: Suggest adding a similar example using {finalsize} with increasing R0:

```suggestion

## Example with `final_size()`

```{r}
r0_low <- 1.3
r0_high <- 3.3

contact_matrix <- matrix(1.0 / 200.0, 2, 2)
demography_vector <- rep(100.0, 2)
psusc <- matrix(1, nrow = 2, ncol = 1)
susc <- psusc

epi_outcome_low <- final_size(
  r0 = r0_low,
  contact_matrix = contact_matrix,
  demography_vector = demography_vector,
  susceptibility = susc,
  p_susceptibility = psusc
)

epi_outcome_high <- final_size(
  r0 = r0_high,
  contact_matrix = contact_matrix,
  demography_vector = demography_vector,
  susceptibility = susc,
  p_susceptibility = psusc
)

expect_true(
  all(epi_outcome_high$p_infected > epi_outcome_low$p_infected)
)
```

```

**Review by pratikunterwegs — COMMENTED** [review]
Hi @Bisaloo, looks good overall, I've just added some suggestions, feel free to take the ones you prefer and which mesh well with other reviews.

File: posts/statistical-correctness/index.qmd [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+``` r
+test_that("Results of gemma2 equal those of GEMMA v 0.97", {
+  expect_equal(Sigma_ee, diag(c(18.559, 12.3672)), tolerance = 0.0001)
+  expect_equal(Sigma_uu, diag(c(82.2973, 41.9238)), tolerance = 0.0001)
+})
+```
+
+Note that even if a **reference** implementation doesn't exist, it is still good practice to compare your implementation to competing ones. Discrepancies might reveal a bug in your implementation or theirs but in any case, finding it out is beneficial to the community.
+
+However, this approach cannot be used in all cases. Indeed, there may not be a reference implementation in your case. Or it might be difficult to replicate identical computations in the case of algorithm with stochasticity [^2].
```
Comment (joshwlambert) [inline posts/statistical-correctness/index.qmd:64]: The `kind` argument in `set.seed()` might help in this situation but I agree that RNG across languages will make replication difficult.

**Comment by joshwlambert** [comment]
As the article is focused on the the correctness of the output of a method, it may be useful to note that this is dependent on valid input and linking to https://blog.r-hub.io/2022/03/10/input-checking/. 

Possible addition, building on Hadley's tweet in the conclusion could be to add something about looking for peer-review of a package, possibly linking to rOpensic (not sure whether this review process covers correctness) or to this blog post: https://notstatschat.rbind.io/2019/02/04/how-do-you-tell-what-packages-to-trust/

File: posts/statistical-correctness/index.qmd (hunk 3 above)
Comment (Bisaloo) [reply posts/statistical-correctness/index.qmd:64]: Even then, I'm not sure it's possible to generate the same random number. In addition with the kind of RNG, you have the initialization problem, and `?set.seed()` states that they have custom initialization methods in some cases. After this, how the `.Random.seed` is used in functions like `random.random()` in python or `runif()` in R may differ.

I would be very interested in hearing if you manage to create the same random value in two languages though :+1:

**Review by pratikunterwegs — APPROVED** [review]
Looks good to me. One caveat: I haven't looked into the files in `_extensions/` or `_freeze`, I assume those are autogenerated in some way or copied from some reference.

**Review by joshwlambert — APPROVED** [review]

**Comment by joshwlambert** [comment]
@Bisaloo all looks good. Nice work. One issue I had was the blog post didn't update for me after I pulled the new changes and ran `quarto preview`, if possible would be good to discuss this.

**Comment by megamezl** [comment]
What other packages might be useful for checking function input, and what are the advantages of using one over the other? does it depend on the goal of the package?

**Comment by Bisaloo** [comment]
@megamezl, on the topic of checking function input, you may like this other blog post: https://blog.r-hub.io/2022/03/10/input-checking/

File: posts/statistical-correctness/index.qmd [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+[^1]: But see the ["testing statistical software" post from Alex Hayes](https://www.alexpghayes.com/post/2019-06-07_testing-statistical-software/) where he presents his process to determine if he deems a statistical package trustworthy or not.
+
+In this blog post, we explore various methods to ensure the statistical correctness of your software. We argue that these tests should be part of your continuous integration system, to ensure your tools remains valid throughout its life, and to let users verify how you validate your package. Finally, we show how these principles are implemented in the Epiverse TRACE tools.
+
+The approaches presented here are non-exclusive and should ideally all be added to your tests. However, they are presented in order of stringency and priority to implement. We also take a example of a function computing the centroid of a list of points to demonstrate how you would integrate the recommendations from this post with the [`{testthat}` R package](https://testthat.r-lib.org/), often used from unit testing:
+
+```{r}
+#' Compute the centroid of a set of points
+#'
+#' @coords Coordinates of the points as a list of vectors. Each element of the
```
Comment (Bisaloo) [inline posts/statistical-correctness/index.qmd:28]:
```suggestion
#' @param coords Coordinates of the points as a list of vectors. Each element of the 
```

**Pull request merged**